### PR TITLE
[ATO-2102] Instrument `ValidationAction.run`

### PR DESCRIPTION
**Pull request by Tawakalt**
**Proposed changes**:
- Fixes [ATO-2102](https://rasahq.atlassian.net/browse/ATO-2102)

View from Jaegar UI:
Screenshot 2024-02-09 at 14 57 00

**Status (please check what you already did)**:
- [X] made PR ready for code review
- [X] added some tests for the functionality
- [ ] updated the documentation in the [rasaHQ/rasa](https://github.com/rasaHQ/rasa)
- [X] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa-sdk/tree/main/changelog) for instructions)
- [X] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa-sdk#code-style) for instructions)

[ATO-2102]: https://rasahq.atlassian.net/browse/ATO-2102?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: tests/test_endpoint.py [hunk 1]
```diff
@@ -23,7 +23,7 @@ def test_server_health_returns_200():
 def test_server_list_actions_returns_200():
     request, response = app.test_client.get("/actions")
     assert response.status == 200
-    assert len(response.json) == 3
+    assert len(response.json) == 4
```
Comment (Tawakalt) [inline tests/test_endpoint.py:26]: We'll have to update this for every subclass of `Action` we create while writing new tests.

File: rasa_sdk/tracing/instrumentation/instrumentation.py [hunk 2]
```diff
@@ -70,9 +71,11 @@ async def async_wrapper(self: T, *args: Any, **kwargs: Any) -> S:
             if attr_extractor and should_extract_args
             else {}
         )
-        with tracer.start_as_current_span(
-            f"{self.__class__.__name__}.{fn.__name__}", attributes=attrs
-        ):
+        if issubclass(self.__class__, ValidationAction):
+            span_name = f"ValidationAction.{self.__class__.__name__}.{fn.__name__}"
```
Comment (Tawakalt) [inline rasa_sdk/tracing/instrumentation/instrumentation.py:75]: Added this cos the class name is the name of the custom action.

File: tests/tracing/instrumentation/test_validation_action.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
+from typing import Sequence, Optional
+
+import pytest
+from opentelemetry.sdk.trace import ReadableSpan, TracerProvider
+from opentelemetry.sdk.trace.export.in_memory_span_exporter import InMemorySpanExporter
+
+from rasa_sdk.tracing.instrumentation import instrumentation
+from tests.tracing.instrumentation.conftest import (
+    MockValidationAction,
+    # MockValidationActionSubClass,
```
Comment (ancalita) [inline tests/tracing/instrumentation/test_validation_action.py:10]: Comment should be removed.

File: tests/tracing/instrumentation/test_validation_action.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            '["name", "address"]',
+        ),
+    ],
+)
+@pytest.mark.asyncio
+async def test_tracing_action_executor_run(
+    tracer_provider: TracerProvider,
+    span_exporter: InMemorySpanExporter,
+    previous_num_captured_spans: int,
+    events: Optional[str],
```
Comment (ancalita) [inline tests/tracing/instrumentation/test_validation_action.py:32]: Isn't the type her supposed to be List[Event]?

File: tests/tracing/instrumentation/test_validation_action.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    expected_slots_to_validate: Optional[str],
```
Comment (ancalita) [inline tests/tracing/instrumentation/test_validation_action.py:33]: And this should always be string, never None?

File: tests/test_endpoint.py (hunk 1 above)
Comment (ancalita) [reply tests/test_endpoint.py:26]: Let's add more detail here about all the action subclasses in the response.

Comment (Tawakalt) [reply tests/test_endpoint.py:26]: more comments or more assert statements?

Comment (ancalita) [reply tests/test_endpoint.py:26]: assertions if possible.

File: tests/test_endpoint.py [hunk 6]
```diff
@@ -23,7 +23,16 @@ def test_server_health_returns_200():
...
+
+    # ENSURE TO UPDATE AS MORE ACTIONS ARE ADDED IN OTHER TESTS
+    expected = [
+        {"name": "custom_async_action"},
+        {"name": "custom_action"},
+        {"name": "custom_action_exception"},
```
Comment (ancalita) [inline tests/test_endpoint.py:32]: I'd leave a top level comment where these actions are defined (in `test_actions.py`) and a reference for the 4th mock validation action.

**Review by ancalita — APPROVED** [review]
🚀

File: tests/test_endpoint.py (hunk 6 above)
Comment (Tawakalt) [reply tests/test_endpoint.py:32]: what's a top level comment? 🙈

Comment (ancalita) [reply tests/test_endpoint.py:32]: Oh I just meant a comment in L29 for all the 3 actions + a 2nd comment above the mock validation action, that's what I meant by top 😄

Comment (Tawakalt) [reply tests/test_endpoint.py:32]: Got it! 😂

**Pull request merged**